main_boaz.tex

```diff
@@ -0,0 +1,910 @@
+
+
+  \item An element of an axis $\name{ax}$ is known as an \emph{index}. We think of an axis $\name{name}$ as describing a \emph{type} that inherits from the type $X$. Therefore, if (for example) $Supp(\name{ax})$ is the integers then we can use operations such as addition, subtraction, multiplication on indices.  Even if two axes have the same support, we still require explicit casting to translate between one to the other, and so (for example) if $i \in \name{batch}$ then it cannot be used as is to axis into a vector of shape $\{ \name{time} \}$. In contrast, if two axes differ only in annotation, then we can translate indices between one and the other without explicit casting. We can think of an annotated axis of the form $\name{ax}_annot$ as a parameter of type $\name{ax}$ and name \texttt{ax\_annot}. (In this sense annotation is like ``Hungarian notation'' for parameter names.) 
+
+
+  
+  \item A \emph{slice} is a subset of an axis. (This does not need to be a strict subset, and so the ax itself is also a valid slice.)  If \name{ax} is an axis and $S \subseteq Supp(\name{ax})$ then  $\name{ax}[S]$ is the slice corresponding to all triples $(ax,annot,x) \in \name{ax}$ such that $x\in S$.  If $Supp(\name{ax})$ is a set of integers, then we use the notation $\name{ax}[..n]$ for the $\name{ax}[\{1,\ldots, n \}]$. We will sometimes use annotation and slicing together and so use the notation such as $\name{layer}_1 = \name{layer}[..128]$ to denote that $\name{layer}_1$ is the annotation of $\name{layer}$ with support $\{1,..,128 \}$.  
+   
+  \item A (tensor) \emph{type} $\mathcal{T}$ is an unordered set of axes. If $\mathcal{T} = \{ \name{ax}_1 ,\ ldots, \name{ax}_d \}$ is a type then a \emph{shape} of type $\mathcal{T}$ is an unordered set of the form $\{ \name{ax}_1[S_1] ,\ldots, \name{ax}_d[S_d] \}$. That is, a shape of type $\mathcal{T}$ is a set of slices of the axes of $\mathcal{T}$. For example, $\{ \name{width} , \name{height}, \name{batch} \}$ is a tensor type while $\{ \name{width}[32], \name{height}[32], \name{batch}[128] \}$ is a shape. Note in both shapes and types, there are no two axes or slices that have both identical names and annotations.
```

Comment:
Is the purpose of the tensor type/shape distinction the need to define a single function that is polymorphic over slices of different sizes? Are there any other purposes?

Reply:
There are two purposes.
The main one is that indeed I think notation is much cleaner if you defined a function that is polymorphic across slices.

Another is that sometimes the length of slices in output could be a dynamic function of the content, and so you could not express it as the type of a function without getting very awkward. 

I think the mechanism of defining a function polymorphically and then specifying pre-conditions and post-conditions on the lengths makes more sense. In a paper these pre and post conditions might be something you ignore in a first reading (and they can be even put in another section).